lighthouse-core/lib/page-functions.js

```diff
@@ -113,18 +113,24 @@ function getElementsInDocument(selector) {
  */
 /* istanbul ignore next */
 function getOuterHTMLSnippet(element, ignoreAttrs = []) {
-  const clone = element.cloneNode();
-
-  ignoreAttrs.forEach(attribute =>{
-    clone.removeAttribute(attribute);
-  });
-
-  const reOpeningTag = /^[\s\S]*?>/;
-  const match = clone.outerHTML.match(reOpeningTag);
+  try {
+    if (ShadowRoot.prototype.isPrototypeOf(element) && element.host && element.localName !== 'a') {
+      element = element.host;
+    }
 
-  return (match && match[0]) || '';
+    const clone = element.cloneNode();
+    ignoreAttrs.forEach(attribute =>{
+      clone.removeAttribute(attribute);
```

Comment:
the types on this whole function get kind of broken if we aren't really accepting just `Element`.

`removeAttribute` and `outerHTML` (and more or less `localName`) exist only on `Element`.

Maybe we should type it as `@param {Element|ShadowRoot} element` and (if we don't need that `localName` check) have the conditional be just `if (element instanceof ShadowRoot) element = element.host;` (with a comment somewhere explaining that this function handles the ShadowRoot case).

It's still playing a little fast and loose with types (it doesn't help that the tsc built in types don't recognize `ShadowRoot` as a global value), but it simplifies things a bit.